### PR TITLE
Add caching, tracing, logging, and batching to the plugin

### LINKED ISSUE #412: Add tracing, logging, and caching to the plugin
### What feature do you want to see added?

In order to measure the performance of the plugin when accessing the database, we should add open telemetry tracing and logging. In addition, there are opportunities to introduce caching so that the plugin doesn't have to access the database each time it needs to access data. instead it can store the data temporarily in a cache to improve access times.

### Upstream changes

_No response_

### Are you interested in contributing this feature?

yes, I plan to create a pull request soon to facilitate these changes.

### PR DESCRIPTION
**Pull request by craigtmoore**
Adds caching using Caffine cache, logging with java.util.logging, tracing using opentelemtry, and inserts the data into the database using batching. We have some builds on our Jenkins instance that have over 29,000 test results so we added these changes to optimize performance. 

Fixes #412

- Update DatabaseTestResultStorage:
  * add caseResultsCache
  * add packageResultsCache
  * update publish() method to use batch updates to improve performance
  * Add @WithSpan annotation to each method
  * add logger
- Update DatabaseTestResultStorageTest:
  * add basic unit-test with a mocked database
  * move duplicate code segments to constants and methods
  * print tables with proper column widths (and truncate values that are too large)
- Update pom.xml:
  * add opentelemtry dependencies (for tracing)
  * add caffine dependency (for caching)
  * add dependency on `kotlin-std-lib-jkd8` to fix dependency resolution issues
  * tie the `hpi` goal to the `compile` phase
- update docker-compose.yaml:
  * add commented out 'mysql' database config
  * add jaeger server
  * add jenkins server with otel agent embedded
- docker-compose-with-zipkin.yaml: same as docker-compose.yaml, but using zipkin instead of jaeger
- Add Dockerfile: creates an instance of the Jenkins docker image that includes the opentelemetry agent jar file, used to generate theweatherman/jenkins:lts-jdk17-otel image.
- Add setup_jenkins.sh: bash script for installing the plugin and configuring jenkins to store junit results in the databse (works with the docker-compose.yaml)

### Testing done

I re-used the existing tests to verify that I've introduced no new regressions, but also addeed a new unit-test to the `getCaseResults_mockDatabase()` method and verify that it loads the data correctly.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [NA] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,14 +22,79 @@ Tables will be automatically created.
...
+Once jenkins is deployed you can configure it using the `./setup_jenkins.sh` shell script provided in the repository.
+
+> **NOTE** you must first login to Jenkins and create an API token, you can do this by going to 
+> `http://localhost:8080/user/admin/configure` and clicking on 'Add new Token' under the 'API Token' section. Then you 
+> must copy the token and pass it as a parameter to the setup script.
+
+Here's an example of how you use the setup script:
+
+```
+./setup_jenkins.sh http://<hostname>:8080 <api_token>
```
Comment (timja) [inline README.md:75]: why not do this in the Dockerfile, 

Something like:
```
# install the plugin from update center to get all dependencies
RUN jenkins-plugin-cli --plugins junit-sql-storage configuration-as-code
# override with locally built version
COPY target/(whatever it is).hpi /usr/share/jenkins/ref/plugins/junit-sql-storage.jpi 

# configuration as code config for configuring this plugin
COPY docker/jenkins.yaml /usr/share/jenkins/ref/jenkins.yaml
```

File: README.md [hunk 2]
```diff
@@ -60,6 +129,52 @@ unclassified:
...
+
+You can access the Jaeger UI by going to `http://localhost:16686`, here you can view the performance of the Jenkins server 
+and your plugin changes.
+
+### Accessing the database
+
+You can also query the mysql (or postgres) database by connecting to the `db` container (note: `winpty` is only required when running on Windows git bash)
+
+```
+$ winpty docker exec -it junit-sql-storage-plugin-db-1 mysql --password=root mysql
```
Comment (timja) [inline README.md:148]: winpty can be mentioned separately, shouldn't be in the default example:
```suggestion
$ docker compose exec db mysql --password=root mysql
```

File: docker-compose-with-zipkin.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+version: '3.1'
+
+services:
+  db:
+    image: mysql:8.0.37-debian
```
Comment (timja) [inline docker-compose-with-zipkin.yaml:5]: I don't think it needs to be this specific:

what about either:
```suggestion
    image: mysql:lts
```

or:
```suggestion
    image: mysql:8.0-debian
```

File: docker-compose-with-zipkin.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    ports:
+      - "3306:3306"
+
+  zipkin:
+    image: openzipkin/zipkin
+    ports:
+      - 9411:9411
+
+  jenkins:
+    image: theweatherman/jenkins:otel
```
Comment (timja) [inline docker-compose-with-zipkin.yaml:18]: Can be use the official Jenkins image please especially since we're already copying in the agent?

File: docker-compose-with-zipkin.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    ports:
```
Comment (timja) [inline docker-compose-with-zipkin.yaml:19]: Can be use the official Jenkins image please especially since we're already copying in the agent?

Shouldn't this have a build statement:

```suggestion
    build:
      context: .
    ports:
```

File: pom.xml [hunk 6]
```diff
@@ -109,22 +114,60 @@
...
+            <artifactId>opentelemetry-sdk-metrics</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-exporter-otlp</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-exporter-logging</artifactId>
+        </dependency>
```
Comment (timja) [inline pom.xml:141]: these imports should all be replaced by a dependency on the opentelemetry plugin so that multiple versions aren't bundled into Jenkins causing problems:
https://plugins.jenkins.io/opentelemetry/dependencies/

File: pom.xml [hunk 7]
```diff
@@ -109,22 +114,60 @@
...
+        <dependency>
+            <groupId>com.github.ben-manes.caffeine</groupId>
+            <artifactId>caffeine</artifactId>
+            <version>3.1.8</version>
```
Comment (timja) [inline pom.xml:145]: replace with api plugin:
```suggestion
            <groupId>io.jenkins.plugins</groupId>
            <artifactId>caffeine-api</artifactId>
```

File: pom.xml [hunk 8]
```diff
@@ -109,22 +114,60 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.jetbrains.kotlin</groupId>
+            <artifactId>kotlin-stdlib-jdk8</artifactId>
+            <version>1.9.10</version>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>5.12.0</version>
```
Comment (timja) [inline pom.xml:155]: version is set by bom in the parent pom: https://github.com/jenkinsci/plugin-pom/blob/master/pom.xml#L177
```suggestion
```

File: pom.xml [hunk 9]
```diff
@@ -137,4 +180,20 @@
...
+        <plugins>
+            <plugin>
+                <groupId>org.jenkins-ci.tools</groupId>
+                <artifactId>maven-hpi-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>hpi</goal>
+                        </goals>
+                        <phase>compile</phase>
```
Comment (timja) [inline pom.xml:193]: why? this doesn't seem right

File: setup_jenkins.sh [hunk 10]
```diff
@@ -0,0 +1,174 @@
+#! /bin/bash
+
+# Colorized output colors
```
Comment (timja) [inline setup_jenkins.sh:3]: I think this file can be completely replaced by just putting the configuration into the Dockerfile as explained in the README comment

**Review by timja — COMMENTED** [review]
How extensively have you tested this? Have you deployed it to your instance with the large number of results?

My main concern is that the optimised database calls for only reading e.g the fail or skipped count are now being in Java so the full amount needs to be read into memory and filtered client side.

Its possible that most of the methods get called when this is used anyway so it may not impact much, or possibly there could either be more caches.

I assume you found a cache was needed because retrieval was slow? do you have any info on this? 

The batch changes make full sense

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+# Download and add the OpenTelemetry Java agent JAR
+ADD https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/latest/download/opentelemetry-javaagent.jar /opt/opentelemetry-javaagent.jar
+
+# Switch to root user and change ownership/permissions of the Java agent JAR
+USER root
+RUN chmod 644 /opt/opentelemetry-javaagent.jar
+RUN chown jenkins:jenkins /opt/opentelemetry-javaagent.jar
+
+USER jenkins
```
[the remark this answers is not in the document]
Comment (craigtmoore) [reply Dockerfile]: Thanks, I didn't know this was added. I've updated the Dockerfile.

File: README.md [hunk 12]
```diff
@@ -22,14 +22,79 @@ Tables will be automatically created.
...
+To try out your chagnes you can deploy Jenkins using the `docker-compose` file provided in the repository.
+```
+$ docker compose up
+```
+This will deploy postgres, jaeger, and Jenkins servers. You'll need to monitor the output of the Jenkins container in 
+order to get the initial admin password.
 
-e.g. you could install the [PostgreSQL Database](https://plugins.jenkins.io/database-postgresql/) plugin.
+```
+$ docker logs -f junit-sql-storage-plugin_jenkins_1
```
[the remark this answers is not in the document]
Comment (craigtmoore) [reply README.md]: Thanks, I've updated it to use `logs` from `docker compose`.

File: README.md [hunk 13]
```diff
@@ -22,14 +22,79 @@ Tables will be automatically created.
 
 ## Getting started
 
-Install your database vendor specific plugin, you can use the Jenkins plugin site to search for it:
+To install the plugin login into Jenkins → Manage Jenkins → Manage Plugins → Available → Search for 'JUnit SQL Storage' → Install.
```
Comment (timja) [inline README.md:25]:
```suggestion
To install the plugin login to Jenkins → Manage Jenkins → Manage Plugins → Available → Search for 'JUnit SQL Storage' → Install.
```

File: src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java [hunk 14]
```diff
@@ -60,202 +78,269 @@
...
+        consoleHandler.setLevel(Level.INFO);
+        logger.addHandler(consoleHandler);
+
+        // Remove default handlers to prevent duplicate messages
+        Logger rootLogger = Logger.getLogger("");
+        Handler[] handlers = rootLogger.getHandlers();
+        for (Handler handler : handlers) {
+            rootLogger.removeHandler(handler);
+        }
+        rootLogger.addHandler(consoleHandler);
```
Comment (timja) [inline src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java:120]: Not sure exactly what you were doing here, seems you're setting logging to INFO?

This would probably simplify what you are trying to do: https://github.com/jenkinsci/jenkins-test-harness/blob/master/src/main/java/org/jvnet/hudson/test/LoggerRule.java

File: src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java [hunk 15]
```diff
@@ -60,202 +78,269 @@
...
+    private void printAndVerifyCaseResultsTable(boolean verifyTable) throws SQLException {
+        try (Connection connection = requireNonNull(GlobalDatabaseConfiguration.get().getDatabase()).getDataSource().getConnection();
+                PreparedStatement statement = connection.prepareStatement("SELECT * FROM caseResults", ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY);
+                ResultSet result = statement.executeQuery()) {
+            if (verifyTable) verifyTableStructure(connection);
+            printResultSet(result);
+        }
+    }
+
+    public void verifyTableStructure(Connection connection)
```
Comment (timja) [inline src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java:254]: why is this needed? (and why public?)

File: README.md (hunk 2 above)
Comment (craigtmoore) [reply README.md:148]: Good point, I've removed it from the example.

File: README.md (hunk 1 above)
Comment (craigtmoore) [reply README.md]: Wow thats great, I didn't know about the `jenkins-plugin-cli` binary or I would have done it this way :-/ I tried it out and it works fine.

File: pom.xml (hunk 6 above)
Comment (craigtmoore) [reply pom.xml:141]: I tried to do as you suggested, but that plugin is not available at
* https://repo.jenkins-ci.org/public/org/jenkins-ci/plugins/

Do you know the artifact details for the maven pom.xml?

File: setup_jenkins.sh (hunk 10 above)
Comment (craigtmoore) [reply setup_jenkins.sh]: I agree, I've removed it.

File: pom.xml (hunk 9 above)
Comment (craigtmoore) [reply pom.xml:193]: When I tried doing `mvn hpi:hpi` for the first time it always failed so I moved it to the compile phase so it would come after the code is compiled.

File: pom.xml (hunk 7 above)
Comment (craigtmoore) [reply pom.xml:145]: Done

File: pom.xml (hunk 8 above)
Comment (craigtmoore) [reply pom.xml]: OK, I've removed it.

File: README.md (hunk 13 above)
Comment (craigtmoore) [reply README.md]: OK, I've re-worded it.

File: src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java (hunk 14 above)
Comment (craigtmoore) [reply src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java]: Yep, that's great. Didn't know about it. I've implemented it.

File: src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java (hunk 15 above)
Comment (craigtmoore) [reply src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java]: I've made the method private (oops).

**Comment by craigtmoore** [comment]
As far as testing goes, the unit-tests at our company are quite large, over 29,000 test results in a single build. We're looking into using a plugin like this, but the performance was quite poor when first tried it (which is why I looked at improving the performance). At first I deployed it using mysql instead of postgres because I could not connect the database to Jenkins.

Once I was able to deploy it using and connect it to a dockerized mysql database, I ran the following pipeline:

```
pipeline {
    agent any
    stages {
        stage('Extract test results') {
            steps {
                sh 'unzip -o /tmp/archive.zip -d build'
            }
        }
    }
    post {
        always {
            junit 'build/**/TEST-*.xml'
        }
    }
}
```

Basically, I downloaded the zip file with the large number of test results from our production Jenkins server and then
copied that to the Jenkins instance, using:
```
docker cp archive.zip junit-sql-storage-plugin-jenkins-1:/tmp/archive.zip
```

Then I ran the pipeline and it took a long time to store the test results (nearly 3 minutes to store the results using
a mysql database). So I added open telemetry to the plugin so I could see where the bottlenecks were. I decided to add
batching to the publish method to help reduce the time it takes to store the values. This certainly helped, but the
other problem was loading the test results took a really long time. I thought it might have been that the queries were
too slow, but it turned out that the junit plugin was calling the `getAllPackageResults` over 1500 times
(which can takes 100s of milliseconds) so I added caching which reduced the processing time down to 10s of microseconds for each method call. I decided to take it one step further and cache the value returned by the `retrieveCaseResults()` method
(which I renamed `getCaseResults()`), this also really helped because the case results query takes ~20ms to run and
that takes a long time when there are over 29000 test results. Also, that query was being called by many of the
 meta-data methods, like:
* getFailedTests
* getSkippedTests
* getPassedTests
By caching the caseResults, we avoided running the query repeatedly and reduced the processing time to 10s of
microseconds.

So to answer your comment, yes I did a lot of testing to verify that my changes improved the performance. The telemetry that I added was also very useful in figuring out the bottle necks.

File: pom.xml (hunk 9 above)
Comment (timja) [reply pom.xml:191]: can you raise an issue at https://github.com/jenkinsci/maven-hpi-plugin/

I don't understand that plugin well enough but don't really want this here, I've never had the need to call `mvn hpi:hpi` directly.

Either:
`mvn clean package` or `mvn hpi:run` should be sufficient and work fine

File: pom.xml (hunk 6 above)
Comment (timja) [reply pom.xml:145]: Its on the page I linked

File: jenkins-config.yaml [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+  globalDatabaseConfiguration:
+    database:
+      postgreSQL:
+        database: postgres
+        hostname: db
+        password: postgres
+        username: postgres
+        validationQuery: "SELECT 1"
+  junitTestResultStorage:
+    storage: "database"
```
Comment (timja) [inline jenkins-config.yaml:11]:
```suggestion
    storage: "database"

```

**Review by timja — COMMENTED** [review]
Getting close

File: README.md [hunk 17]
```diff
@@ -60,6 +114,52 @@ unclassified:
     storage: "database"
 ```
 
+Here's an example of how to use it
+
+```
+java -jar java-cli.jar -s http://localhost:8080 -u admin -p <api_token> -f junit-sql-storage-plugin-config.yml
```
[the remark this answers is not in the document]
Comment (craigtmoore) [reply README.md]: Good point, I've updated it. It was completely wrong :-P

File: README.md [hunk 18]
```diff
@@ -60,6 +114,52 @@ unclassified:
...
+```
+
+### Accessing Jaeger
+
+You can access the Jaeger UI by going to `http://localhost:16686`, here you can view the performance of the Jenkins 
+server and your plugin changes.
+
+### Accessing the database
+
+You can also query the mysql (or postgres) database by connecting to the `db` container.
```
[the remark this answers is not in the document]
Comment (craigtmoore) [reply README.md:115]: OK

File: deploy.sh [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+  exit 1
+fi
+echo
+# Build Docker Image
+echo -e "${LightGreen}docker build --tag jenkins .${NC}"
+docker build --tag jenkins .
+echo
+# Deploy the docker swarm
+echo -e "${LightGreen}docker compose up -d${NC}"
+docker compose up -d
```
[the remark this answers is not in the document]
Comment (craigtmoore) [reply deploy.sh:23]: I couldn't get that to work properly. It didn't build the Docker image so I've kept it as a 2 step operation.

File: docker-compose-with-zipkin.yaml (hunk 3 above)
Comment (craigtmoore) [reply docker-compose-with-zipkin.yaml:5]: I think I'll just remove that file. It doesn't really add anything.

File: docker-compose-with-zipkin.yaml (hunk 4 above)
Comment (craigtmoore) [reply docker-compose-with-zipkin.yaml]: I've removed this file so I guess this topic is moot.

File: jenkins-config.yaml (hunk 16 above)
Comment (craigtmoore) [reply jenkins-config.yaml]: I've added a line break to the end of the file.

File: docker-compose.yaml [hunk 20]
```diff
@@ -8,3 +8,25 @@ services:
...
+      - "8080:8080"
+      - "50000:50000"
+    environment:
+      OTEL_EXPORTER_ZIPKIN_ENDPOINT: http://jaeger:9411/api/v2/spans
+      OTEL_TRACES_EXPORTER: zipkin
+      OTEL_SERVICE_NAME: jenkins
+      OTEL_METRICS_EXPORTER: none
+      OTEL_LOGS_EXPORTER: none
+      OTEL_JAVAAGENT_LOGGING: simple
+      JAVA_OPTS: "-javaagent:/opt/opentelemetry-javaagent.jar"
```
[the remark this answers is not in the document]
Comment (craigtmoore) [reply docker-compose.yaml]: Sounds good. I've remove the opentelemetry-javaagent and added the setup wizard skip property

File: pom.xml (hunk 6 above)
Comment (craigtmoore) [reply pom.xml]: Thanks, I was able to make it work, but I had to replace all the @WithSpan annotations that I had in the Java code and use manual tracing instead 🤷‍♂️

File: pom.xml (hunk 9 above)
Comment (craigtmoore) [reply pom.xml]: OK, I've removed that section from the pom.xml and adjusted the ./deploy.sh file.

**Comment by craigtmoore** [comment]
I think I've resolved all of your comments, please let me know if there is any thing else. I really appreciate all of the feedback. I'm going to do a bit more testing, especially with junit attachments plugin.

**Review by timja — COMMENTED** [review]
This looks really good, I've tested it out, fixed up the deploy script to work with build and a couple of other minor things.

I've found one issue though with the open telemetry integration, I would expect the spans to be linked to the build but they aren't:

image

I'm wondering if the `junit` step needs a `StepHandler`?

https://github.com/jenkinsci/opentelemetry-plugin/blob/03559ec4adf9ed6843e7fb0635a5e4cc9f961fb2/src/main/java/io/jenkins/plugins/opentelemetry/job/step/DurableTaskHandler.java

see also:
https://github.com/jenkinsci/opentelemetry-plugin/issues/850

I couldn't see any `TRACEPARENT` env values being set

cc @cyrille-leclerc or @kuisathaverat if you can provide any advice

File: src/main/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorage.java [hunk 21]
```diff
@@ -114,60 +156,96 @@ private static class RemotePublisherImpl implements RemotePublisher {
...
-                            } else {
-                                statement.setNull(11, Types.VARCHAR);
-                            }
-                            if (StringUtils.isNotEmpty(caseResult.getErrorStackTrace())) {
-                                statement.setString(12, StringUtils.truncate(caseResult.getErrorStackTrace(), MAX_STACK_TRACE_LENGTH));
-                            } else {
-                                statement.setNull(12, Types.VARCHAR);
+        @Override
+        public void publish(TestResult result, TaskListener listener) throws IOException {
+            var publishSpan = createSpan("DatabaseTestResultStorage.RemotePublisherImpl.publish");
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorage.java:161]: As far as I can tell this only works for agents on the controller not remote agents

Comment (timja) [reply src/main/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorage.java:161]: Setting `-Dotel.java.global-autoconfigure.enabled=true` on the agent seems to work

**Comment by timja** [comment]
I don't think its `StepHandler` I think if the variables are propagated through createRemotePublisher and this code is copied then it would work:
https://github.com/jenkinsci/opentelemetry-plugin/blob/82e3e1b2574a68864dcd05b43b95d0aeb7f41249/src/main/java/io/jenkins/plugins/opentelemetry/job/OtelEnvironmentContributorService.java#L41-L52

**Comment by timja** [comment]
Continued at https://github.com/jenkinsci/junit-sql-storage-plugin/pull/414

**Comment by jonesbusy** [comment]
Shoudn't the opentelemetry plugin be optional ? What about jenkins instance using SQL storage but not connected to an Opentelemetry collector ?

From what I remember few month ago is that having OpenTelemetry plugin installed will display some deak links on job/build page when the plugin is not configured

**Comment by timja** [comment]
Possibly, would be better to remove those dead links I think if the plugin isn't configured at all.

Plugins should be able to integrate without it adding a bunch of things to the UI

**Comment by cyrille-leclerc** [comment]
Thanks for your interest in the opentelemetry plugin.
I think we could make it optional for simpler integration of otel in the other Jenkins plugins, I have a few checks to do. 
I'll get back to you ASAP

**Comment by timja** [comment]
just raised a PR https://github.com/jenkinsci/opentelemetry-plugin/pull/868

**Comment by kuisathaverat** [comment]
> I couldn't see any TRACEPARENT env values being set

you have to enable the export of the environment variables that define the TRACEPARENT, but I not sure how this plugin get the Opentelemetry context, it probably has to retrieve it from the job not front and environment variable. We store the context in actions in the job.

**Comment by timja** [comment]
I was able to get it in https://github.com/jenkinsci/junit-sql-storage-plugin/pull/414

**Comment by cyrille-leclerc** [comment]
@craigtmoore @timja the way you integrate OpenTelemetry is great. `GlobalOpenTelemetry.get()` is the way to go.
I'm looking at some improvements to this pattern `GlobalOpenTelemetry.get()`. I'll get back to you shortly.

I saw @timja' PR, Ill review it it asap, I have to understand the challenge you faced :-)
* https://github.com/jenkinsci/opentelemetry-plugin/pull/868

**Comment by timja** [comment]
Anything running on the controller I think would be quite straightforward, but running on agents is a bit more complicated, its not the nicest / cleanest implementation in https://github.com/jenkinsci/junit-sql-storage-plugin/pull/414
